### ISSUE TITLE
Coherence ignores important checks

### ISSUE THREAD

**Issue by VictorTaelin** [VictorTaelin/Interaction-Type-Theory#1]
I believe Coherence may not be sufficient for our needs. The first issue is that it's still unclear how to check predicates. According to coherence, `P x` and `P y` will always be equal, with the arguments being ignored. Also, when we have kinds (like, `x : P : *`), coherence will ignore P and check as though it was `x : *`.

![img0](https://i.imgur.com/sbuYKr9.jpg)
![img1](https://i.imgur.com/cxzjO6K.jpg)

It seems what we actually need is to navigate the graph using the λ-calculus readback strategy. I'm not a huge fan of this, since ports 1 and 2 of the `CON` nodes are asymmetric (port 1 represents a variable, port 2 represents an application). However, this does make sense if our goal is to use nets to represent λ-calculus proofs. It's a bit underwhelming though, since coherence is exponentially faster than λ-calculus readback. For example, coherence checks equality for the church-encoded `2^N` in linear time, rather than exponential time. Indeed, we can easily perform church-encoded equality in linear time on interaction combinators. As such, it makes sense that our type checker's definitional equality should be able to do the same. So, it still seems like something is missing. Perhaps there is a way to avoid exponential blowup on λ-calculus readback?

On a brighter side, this example made me realize that application nodes and arrow nodes are "overloaded", i.e., the same node that is an application (on the value level) can be represented as an arrow (on the type level). Specifically, the `λP` node is used as both the `(f P)` application, and as the `(T -> (P -> P))` arrow. It also clarifies what to do when we reach an application node on its port 1; it is interpreted as a `∀` binder. This is a good aspect as it allows `P == P` to work, even if the left and right P come from different nodes (i.e., the `->` binder and the actual λ). Basically, the process becomes a Bruijn index.

**Comment by VictorTaelin** [comment]
Pasting from Discord, for the record:

read all that you said. the way I got coherence was:

1. if you readback net to a λ-term, the behavior of ANN is that you just passthrough from port 1 to 2, as if it was a wire. then, you annotate it with the type coming from port 0.

2. because of that, when you pass through TWO consecutive ANN nodes, you essentially annotate a term twice, like: `x : Bool : Bool`

3. since a term can only have one type, it must follow that `Bool == Bool`. thus, in a way, the function of ANNs is to "set up equality points", i.e., ports you must check for equality. that's all ANN does

4. now, the question is: what is the meaning of equality? how do we check if two pointers are equal?

5. the solution, since we want to avoid λ-calculus dependency, I came up with a notion of "con equivalence", which is like Mazza's equivalence, except DUP-invariant. that is the only way to make it work, otherwise it would be impossible to encode even simple types like `∀(x: P) -> P -> P`, because each `P` would be considered to be different under Mazza's original equivalence!

6. turns out that this equivalence can be implemented efficiently via the execution paths algorithm. we can compare church encodings for equality in logarithmic time (like, `2^256 == 2^128*2^128`).

7. with ANNs and equality, all that I had to do is detect paired ANNs (in order to call equal() on the correct wires). after thinking a little bit, I noticed a clever way to do it. by just considering all possible paths on the net (from root to root), which I call a leap, and checking if they're symmetric (stack == queue), we are kind-of checking equality between the downwards path and the upwards path. now, if we only consider paths that are ANN-symmetric, this will discard any path that doesn't traverse two nodes. so, essentially, the effect is that coherence automagically coincides with checking con-equivalence of the port 0 of all paired ANN nodes. and that's really cool

8. this sounded all too good, but... sadly, con-equivalence seems insufficient for a traditional type theory. first, when we're dealing with a non-closed net, such as checking `(f x) == (f y)`, this equality will pass through the APP node directly towards root, ignoring the argument. so, `(f x) == (f y)` merely checks if `f == f`, which is NOT what we want if we're building a conventional type theory. then, when you have chained annotations, like `(x : (A : *) : (B : *))`, coherence will pass through the mid nodes and NOT check `A == B`. it will only check `* == *`. which is not sufficient...

9. so, ultimately, I don't think con-equivalence is sufficient to build a "familiar" type theory, because we absolutely need `f 0` to be different from `f 1`. it under-checks the term. so, I if we want a CoC-like theory, we need actual λ-calculus readback semantics, as it deals with these 2 issues, at the cost of an exponential slowdown in some cases, since all sharing is un-done when checking for equality. perhaps this "coherence type theory" makes sense in some weird mathematical way, but it doesn't seem to lead into a type theory that is familiar to us

**Comment by VictorTaelin** [comment]
I think there are 2 ways to go, though; 1. a con-equivalence / coherence based system, that may be some kind of interesting ultra lightweight alien type theory, but definitely not familiar nor capable of replicating the same things that CoC is doing; 2. a λ-readback based equality, which should give us a lot of CoC analogues. Since the purpose of this is to replace Kind, chances are I'll explore the direction 2 first, but 1 seems extremely appealing for its own sake, it could have applications to AI, program synthesis, etc.

Since it is likely that I will remove the current readme with the coherence condition, here is a snapshot: https://gist.github.com/VictorTaelin/2c8eb3690f6d94b1e05cec2330b8ab9b

**Comment by VictorTaelin** [comment]
I think this can be solved by slightly altering the coherence condition to allow routes that traverse sideways on ANN nodes. The problem is the current algorithm, as written, is ignoring this path:

![path](https://i.imgur.com/kdi0lKt.png)

But it must be considered.

**Comment by VictorTaelin** [comment]
So, after thinking a lot, these examples convinced me λ-readback-equality is necessary. It works exactly as expected in all cases (obviously) and I can't think of a faster or clever way to make these work.

![](https://i.imgur.com/wHbrUTZ.jpg)

The first is `λ (P: Bool -> *) -> λ (t: P True) -> λ (f: P False) -> (f :: P True)` (accidentally inverted the T/F on the image though), which should fail to type check. By running through the net, starting from both paired ANN nodes, towards their port 0, we reach different nodes (False and True) in the highlighted paths, resulting in a type error, as expected.

The second is the result of `λ T . (λx. x) :: ∀(P: T) -> P -> P`, which normalizes to `λ T . λ (f: ∀(P: T) -> P -> P) . λ (P: T) . λ (u: P) -> (u: P)`, which is well-typed. The highlighted path checks the returned value, `u`, which results in an equality between the two highlighted paired ANN nodes, which, when readback become both `P`; thus, `P == P`, and this path checks. Notice the `P` variable is obtained by reaching the port 1 of a CON node, and computing its Bruijn index. The coherence condition, even if it went that far, would return false here, because these `P`'s are coming from different nodes, and would have different CON paths.

So, in short, I believe the way to go is to use λ-calculus readback for equality. It works and makes the most logical sense.

**Comment by VictorTaelin** [comment]
After spending a long time trying to implement λ-calculus readback based equality, I couldn't make it work without tons and tons of hack - which is a big red flag. On the other hands, the current equality algorithm is extremely elegant and logical, but it has one issue: on the `f x == g y` case, it will check `f == g`, but not `x == y`. Which is obviously a problem.

Over the last few days, my mindset was to modify the equality algorithm in order to make this `x == y` check happen; i.e., to align it with λ-calculus readback. To do so, we would need to treat CON nodes differently from all other nodes (which is, itself, suspicious), in a way that entering a CON node through its port 2 would move towards the port 1 (rather than just the port 0, as happens in all other nodes type). Turns out this is not only extremely inelegant, weird and out-of-place, but it brings all sorts of bugs and problems, which I won't fully elaborate now. In short, it feels extremely unlikely that λ-calculus readback equality is the right approach, and the elegant equality algorithm, based on paths equivalence, seems to be the way to go.

But how do we deal with `f x == g y` not checking the arguments? That is absolutely a deal-breaker, right? Well, actually, if we take a moment to think, the original equality algorithm **is actually capable of checking `f x == g y` properly**. For example, on `λf. λx. λy. (f x) == λf. λx. λy. (f y)`, it will actually detect these as different terms. It already works as expected, respecting λ-calculus semantics, on closed terms. So, what is the issue?

The thing is, on my encoding of `(λP. λt. λf. f) :: ∀P. P T -> P F -> P F`, when we normalize that, it will result in a net where ANN nodes only point to subsets of the graph, in a way that excludes the `T == F` check. So, it would seem like we what we need is for the equality algorithm to move towards `T == F` in that specific location. But is it? To think about it, imagine the normal form of that net was the one I drew on the right, rather than the one on the left:

![](https://i.imgur.com/Puxnan9.png)

See, it has additional ANN nodes, which cover the missing `T == F` check. Not by modifying the equality algorithm, but by modifying the net itself. And that actually makes a lot of sense. But this brings up the following question: would it be possible to encode function types and applications differently, so that the encoding of  `(λP. λt. λf. f) :: ∀P. P T -> P F -> P F` reduces to the net on the right (i.e., the one that does check `T == F`)? If yes, then I believe all our problems would be solved. We just update the encodings and leave the current, efficient, elegant and seemingly correct equality algorithm as is. But how? Due to some family health circumstances I'm not in a good mood to work on that anymore, so I'll leave it as is for a while. I wish there was someone capable of helping with that specific question though~

**Comment by VictorTaelin** [comment]
We can indeed force the checking of `x == y` by modifying the encoding of arrows! Here is an example solution:

![](https://i.imgur.com/vGUVZt2.png)

![](https://i.imgur.com/VyXY2sR.png)

![](https://i.imgur.com/3uZ2z2H.png)

This suggests that we indeed got the correct equality algorithm, and that the issue is on the encoding of the function type when translating from CoC to ITT. I'll keep working on it.

**Comment by VictorTaelin** [comment]
We can also enforce the correct checking by just using 2 additional superposition nodes on the encoding of `->`:

![](https://i.imgur.com/nLOCXKB.jpg)

And unlike the approach above, this has the correct readback, rather than resulting in a deadlock. It also does all the side checks in a way that makes sense (like cloning T, F, T to perform the `True : Bool`, `False : Bool`, `True : Bool` checks). Seems like this is the key insight I was missing so far.

**Comment by VictorTaelin** [comment]
tired, so lazily copying from Discord:

so I think I have a solution for everything now (although I didn't have time to test), including several problems I didn't talk about (like ANN nodes connected to true/false would just vanish, checking something that has a kind (i.e., `val : (typ : kind)`) didn't work, finding paired ANNs was complex, weird code to deal with ANN stacks differently, and several others). I won't explain in details rn as it needs testing, but key insights are:

we encode arrows using an additional sup node, as I commented earlier.

equality is actually not related to ANNs, but a new "ball" node with 2 main ports. when it interacts with anything, it just propagates. when it interacts with itself, it becomes a wire, and triggers equality, directed at both sides of this wire

ANN nodes are still used but aren't special. there is no "decay". instead, we just encode `*` (the type of types) as an "identity ANN node". so when we apply functions like `λ(P:*). λ(t: P). λ(f: P). (f: P)` to `*`, the effect is that it will undo the ANNs inside, "compiling" the program to `λt. λt. f`

the ball in action:

![](https://i.imgur.com/bkH0fVA.jpg)

notice the ball starts alone and conveniently becomes a pair exactly on the port 2 of the ANN node, which is the precise location where we need to trigger equality 

ball interactions:

![](https://i.imgur.com/k6WAN44.jpg)

complete type-checking, equality and compilation example:

![](https://i.imgur.com/97UHpUh.jpg)

![](https://i.imgur.com/i0v25WQ.jpg)

should be A=A*

so, doesn't this make a lot of sense? `*` from CoC is just an identity ANN, and when we apply a typed term to it, it removes its associated annotations

on the case below:

![](https://i.imgur.com/RjDi5y0.png)

the `*` causes that nested annotation to become a wire, removing the nested annotation problem. on the other side, the Bool annotation on the left side of the image will cause true, false, true to be checked as booleans - and, once again, the nested annotation vanishes, because Bool is itself a bunch of triangles which will pass through the ANN node, "decorating" its associated booleans. then, on the right side, the ball will cause the other copies of false, true to be checked for equality, performing the `P false == P true` check, which we previously missed. the balls will make sure this check happens... perviously, `F == T` would only be compared when they were bound variables, but not when they were λ-encoded values, as they'd pass through the ANNs and "vanish" - the check would never happen. all this is incredibly simple, beautiful and obvious in retrospect

**Comment by VictorTaelin** [comment]
Today I'm trying to find an encoding of `->` that doesn't accumulate a lot of superpositions.

![](https://i.imgur.com/yxHBa1n.jpg)

> notice all these layered ||'s, they're the dashed trigs (sups) that accumulate. we can just pass through, erasing the right side... as if the top net was connected to an infinite cascade of dashed-trigs-with-right-side-erased... but... ugh

> but at least all the important checks are there, and everything is correct, as if the program tracks its complete checking history, which is kinda nice

> but if we can find an encoding for -> that doesn't accumulate trigs like that, it would be cleaner

> notice there are 3 copies of (true,true). 1 is used for checking P T == P T, one is used for checking T : Bool, T : Bool, and the last one... is just there, wasting space???

To do so, I need to find a solution for this:

![](https://i.imgur.com/nF8cniB.jpg)

Working on that right now...

**Comment by VictorTaelin** [comment]
i'm stupid

![](https://i.imgur.com/HBENCFB.jpg)

good night

**Comment by VictorTaelin** [comment]
im literally stupid

**Comment by VictorTaelin** [comment]
![](https://i.imgur.com/yBta0Rm.jpg)

![](https://i.imgur.com/SYq88sP.jpg)

![](https://i.imgur.com/4nm1QRp.jpg)

**Comment by VictorTaelin** [comment]
relevant example on how to do it: https://twitter.com/VictorTaelin/status/1685032556041736194

**Comment by VictorTaelin** [comment]
Since I forgot, here is the actual initial state of the check above:

![](https://pbs.twimg.com/media/F2KBjaVWgAEqpDb?format=jpg&name=small)

**Comment by VictorTaelin** [comment]
With:

1. Multi-port nodes 

2. Franchu's equality node

3. Free vars

We can develop an interaction type checker that outputs `⊤` for well-typed, and `⊥` for ill-typed, terms, by just evaluating them.

Example:

![](https://i.imgur.com/frQJipY.png)

![](https://i.imgur.com/jmBSGDS.png)

![](https://i.imgur.com/Q0B7bEk.png)

![](https://i.imgur.com/3J3U2xd.png)

**Comment by hydrastro** [comment]
Awesome content, thanks!!!
